File: app/src/main/kotlin/com/openclaw/ghostcrab/domain/model/DiscoveredGateway.kt
package com.openclaw.ghostcrab.domain.model

/**
 * A gateway discovered via mDNS (NsdManager).
 *
 * Service type: `_openclaw-gw._tcp.` Default port: 18789.
 * Deduplication key: [instanceName].
 *
 * @param instanceName mDNS service instance name (unique per gateway on the network).
 * @param hostAddress Resolved IPv4/IPv6 address.
 * @param port TCP port the gateway is listening on.
 * @param displayName Human-readable name from TXT record or instance name fallback.
 * @param version Gateway version from TXT record, or `null` if not advertised.
 */
data class DiscoveredGateway(
    val instanceName: String,
    val hostAddress: String,
    val port: Int,
    val displayName: String,
    val version: String?,
) {
    /** Constructed base URL for use with [com.openclaw.ghostcrab.domain.repository.GatewayConnectionManager]. */
    val url: String get() = "http://$hostAddress:$port"
}

File: ios/GhostCrab/Domain/Model/DiscoveredGateway.swift
import Foundation

/// A gateway discovered via mDNS (NWBrowser / Bonjour).
///
/// Service type: `_openclaw-gw._tcp.` Default port: 18789.
/// Deduplication key: `instanceName`.
///
/// - Parameters:
///   - instanceName: mDNS service instance name (unique per gateway on the network).
///   - hostAddress: Resolved IPv4/IPv6 address.
///   - port: TCP port the gateway is listening on.
///   - displayName: Human-readable name from TXT record or instance name fallback.
///   - version: Gateway version from TXT record, or `nil` if not advertised.
public struct DiscoveredGateway: Codable, Sendable, Hashable {
    public let instanceName: String
    public let hostAddress: String
    public let port: Int
    public let displayName: String
    public let version: String?

    public init(
        instanceName: String,
        hostAddress: String,
        port: Int,
        displayName: String,
        version: String?
    ) {
        self.instanceName = instanceName
        self.hostAddress = hostAddress
        self.port = port
        self.displayName = displayName
        self.version = version
    }

    /// Constructed base URL for use with `GatewayConnectionManager`.
    public var url: String { "http://\(hostAddress):\(port)" }
}
